File: content/problems/2115-find-all-possible-recipes-from-given-supplies/solutions/availability-topological-sort/solution.cpp
class Solution {
    public:vector<string> findAllRecipes(vector<string>& recipes,vector<vector<string>>& ingredients,vector<string>& supplies){
        unordered_map<string,vector<int>> dependents;
        vector<int> missing(recipes.size());
        for(int i=0;i<(int)recipes.size();++i){
            missing[i]=ingredients[i].size();
            for(string& ingredient:ingredients[i])dependents[ingredient].push_back(i);
        }
        queue<string> available;
        for(string& supply:supplies)available.push(supply);
        vector<string> answer;
        while(!available.empty()){
            string item=available.front();
            available.pop();
            for(int recipe:dependents[item])if(!--missing[recipe]){
                answer.push_back(recipes[recipe]);
                available.push(recipes[recipe]);
            }
        }
        return answer;
    }
};

File: content/problems/2115-find-all-possible-recipes-from-given-supplies/solutions/availability-topological-sort/Solution.java
class Solution {
    public List<String> findAllRecipes(String[] recipes,List<List<String>> ingredients,String[] supplies){
        Map<String,List<Integer>> dependents=new HashMap<>();
        int[] missing=new int[recipes.length];
        for(int i=0;i<recipes.length;i++){
            missing[i]=ingredients.get(i).size();
            for(String ingredient:ingredients.get(i))dependents.computeIfAbsent(ingredient,ignored->new ArrayList<>()).add(i);
        }
        Queue<String> queue=new ArrayDeque<>(Arrays.asList(supplies));
        List<String> answer=new ArrayList<>();
        while(!queue.isEmpty()){
            String item=queue.poll();
            for(int recipe:dependents.getOrDefault(item,Collections.emptyList()))if(--missing[recipe]==0){
                answer.add(recipes[recipe]);
                queue.offer(recipes[recipe]);
            }
        }
        return answer;
    }
}
